server/src/services/apps.service.ts

```diff
@@ -74,6 +76,86 @@ export class AppsService {
     return app;
   }
 
+  async clone(existingApp: App, user: User): Promise<App> {
+    let clonedApp: App;
+
+    await this.entityManager.transaction(async (manager) => {
+      clonedApp = await this.createClonedAppForUser(manager, existingApp, user);
+      await this.buildClonedAppAssociations(manager, clonedApp, existingApp);
+    });
+
+    return clonedApp;
+  }
+
+  async createClonedAppForUser(
+    manager: EntityManager,
+    existingApp: App,
+    currentUser: User,
+  ): Promise<App> {
+    const newApp = manager.create(App, {
+      name: existingApp.name,
+      organizationId: currentUser.organizationId,
+      user: currentUser,
+    });
+    await manager.save(newApp);
+
+    const newAppUser = manager.create(AppUser, {
+      app: newApp,
+      user: currentUser,
+      role: 'admin',
+    });
+    await manager.save(newAppUser);
+    return newApp;
+  }
+
+  async buildClonedAppAssociations(manager, newApp: App, existingApp: App) {
+    const dataSourceMapping = {};
+    const newDefinition = existingApp.editingVersion?.definition;
+
+    const existingDataSources = await manager.find(DataSource, {
+      app: existingApp,
+    });
+
+    for (const source of existingDataSources) {
+      const newSource = manager.create(DataSource, {
+        app: newApp,
+        name: source.name,
+        options: source.options,
```

Comment:
If the data source has encrypted credentials, we should clone those too instead of making reference to the same credential entity and then will have to update the options.